rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+implementation detail and not part of our public 'API'?
+
+### instanceUUID
+
+The same instance UUID (`server.uuid` / `{dataFolder}/uuid`) is currently used by all the workers. 
+
+We still need to identify with the teams if this is going to be a problem, in which case we would need to have distinct 
+instance uuid per workers.
+
+Note that this could be a breaking change, as the single `server.uuid` configuration property would not be enough. 
+Could the new `clustering.getWorkerId()` API be used here somehow? We could have a unique worker id with `serverUUid-workerId`. 
+
+### The Dev CLI
+
+In development mode, we are already spawning processes from the main process: The Kibana server running in the main process 
+actually just kickstarts core to get the configuration and services required to instantiate the `CliDevMode` 
+(from the legacy service), which itself spawns a child process to run the actual Kibana server.
+
+Even if not technically blocking, it would greatly simplify parts of the workers and coordinator logic if we were able 
+to finally extract the dev cli from Core and to avoid having it create a temporary server just to access the Legacy service
+and configuration.
+
+Note that extracting and refactoring the dev cli is going to be required anyway to remove the last bits of legacy, 
+as it currently relies on the legacy configuration to run (and is running from the legacy service).
+
+## Technical impact on our plugins codebase
+
+### Identifying things that may break
+
+- Concurrent access to the same resources
+
+Is there, for example, some part of the code that is accessing and writing files from the data folder (or anywhere else) 
+and makes the assumption that it is the sole process actually writing to that file?
+
+- Using instanceUUID as a unique Kibana process identifier
+
+Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
+Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?
+
+- Things needing to run only once per Kibana instance
+
+Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
+or starting schedulers? 
+
+An example would be reporting's queueFactory pooling. As we want to only be running a single headless at a time per 
+Kibana instance, only one worker should have pooling enabled.
+
+### Identified required changes
+
+(Probably not exhaustive at the moment)
+
+#### Reporting
+
+We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
+in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 
+to only have the 'main' worker be pooling for reporting tasks.
+
+#### Telemetry
```

Comment:
Event-based telemetry may be affected as well. Both: the existing one in the Security Solutions team (cc @stevewritescode), and the general one that is in the works.

More specifically, the size of the queues will be multiplied per worker, also growing in the amount of network bandwidth used, and potentially affecting our customers.

We could make sure that the queues are held only in the `main` node.